File: src-api/src/extensions/agent/claude/tools/recognize-image.ts
/**
 * Image Recognition Tool
 *
 * Allows the agent to read and analyze local image files using the configured LLM provider.
 */

import { tool } from '@anthropic-ai/claude-agent-sdk';
import { z } from 'zod';
import { readFile, access, stat } from 'fs/promises';
import { extname } from 'path';

/**
 * Media type mapping for supported image formats
 */
const MEDIA_TYPE_MAP: Record<string, string> = {
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.webp': 'image/webp',
};

/**
 * Maximum file size (5MB)
 */
const MAX_FILE_SIZE = 5 * 1024 * 1024;

/**
 * Supported image extensions
 */
const SUPPORTED_EXTENSIONS = ['.png', '.jpg', '.jpeg', '.gif', '.webp'];

/**
 * Create the RecognizeImage tool
 *
 * @param apiKey - API key for the LLM provider
 * @param baseUrl - Base URL for the API (optional)
 * @param model - Model to use for recognition (optional)
 */
export function createRecognizeImageTool(
  apiKey: string,
  baseUrl?: string,
  model?: string
) {
  return tool(
    'RecognizeImage',
    `识别和分析图片内容。当需要查看、理解或分析图片文件时使用此工具。

支持的格式: PNG, JPG, JPEG, GIF, WEBP
最大文件大小: 5MB

使用场景:
- 识别验证码
- 分析截图内容
- 理解图表和数据可视化
- 提取图片中的文字
- 描述图片内容`,
    {
      imagePath: z
        .string()
        .describe('图片文件的绝对路径,例如: /path/to/image.png'),
      question: z
        .string()
        .optional()
        .describe(
          '关于图片的问题或需要分析的内容(可选)。例如: "请识别验证码", "请描述图片内容", "请提取图片中的文字"。默认为"请描述这张图片的内容"'
        ),
    },
    async (args) => {
      try {
        // 1. Check if file exists
        try {
          await access(args.imagePath);
        } catch {
          return {
            content: [
              {
                type: 'text' as const,
                text: `文件不存在: ${args.imagePath}`,
              },
            ],
            isError: true,
          };
        }

        // 2. Check file size
        const stats = await stat(args.imagePath);
        if (stats.size > MAX_FILE_SIZE) {
          return {
            content: [
              {
                type: 'text' as const,
                text: `图片文件过大 (${(stats.size / 1024 / 1024).toFixed(2)}MB),超过 5MB 限制。请使用较小的图片。`,
              },
            ],
            isError: true,
          };
        }

        // 3. Check file format
        const ext = extname(args.imagePath).toLowerCase();
        if (!SUPPORTED_EXTENSIONS.includes(ext)) {
          return {
            content: [
              {
                type: 'text' as const,
                text: `不支持的图片格式: ${ext}。支持的格式: ${SUPPORTED_EXTENSIONS.join(', ')}`,
              },
            ],
            isError: true,
          };
        }

        // 4. Read and encode image
        const imageData = await readFile(args.imagePath);
        const base64Image = imageData.toString('base64');
        const mediaType = MEDIA_TYPE_MAP[ext] || 'image/png';

        // 5. Prepare API request
        const apiBaseUrl = baseUrl || 'https://api.anthropic.com';
        const apiModel = model || 'claude-sonnet-4-20250514';
        const question = args.question || '请描述这张图片的内容';

        console.log('[RecognizeImage] Analyzing image:', {
          path: args.imagePath,
          size: `${(stats.size / 1024).toFixed(2)}KB`,
          format: ext,
          question: question.slice(0, 50),
        });

        // 6. Call LLM API
        const response = await fetch(`${apiBaseUrl}/v1/messages`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'x-api-key': apiKey,
            'anthropic-version': '2023-06-01',
          },
          body: JSON.stringify({
            model: apiModel,
            max_tokens: 2048,
            messages: [
              {
                role: 'user',
                content: [
                  {
                    type: 'image',
                    source: {
                      type: 'base64',
                      media_type: mediaType,
                      data: base64Image,
                    },
                  },
                  {
                    type: 'text',
                    text: question,
                  },
                ],
              },
            ],
          }),
        });

        if (!response.ok) {
          const errorText = await response.text();
          console.error('[RecognizeImage] API error:', {
            status: response.status,
            error: errorText,
          });
          return {
            content: [
              {
                type: 'text' as const,
                text: `API 调用失败 (HTTP ${response.status}): ${errorText}`,
              },
            ],
            isError: true,
          };
        }

        // 7. Parse response
        const result = (await response.json()) as {
          content?: Array<{ type: string; text?: string }>;
          error?: { message: string };
        };

        if (result.error) {
          return {
            content: [
              {
                type: 'text' as const,
                text: `识别失败: ${result.error.message}`,
              },
            ],
            isError: true,
          };
        }

        const text =
          result.content?.find((c) => c.type === 'text')?.text ||
          '无法识别图片内容';

        console.log('[RecognizeImage] Recognition successful:', {
          resultLength: text.length,
        });

        return {
          content: [
            {
              type: 'text' as const,
              text,
            },
          ],
        };
      } catch (error) {
        console.error('[RecognizeImage] Error:', error);
        return {
          content: [
            {
              type: 'text' as const,
              text: `图片识别失败: ${error instanceof Error ? error.message : String(error)}`,
            },
          ],
          isError: true,
        };
      }
    }
  );
}

File: src-python/core/agent/tools/recognize_image.py
"""Image recognition tool for Claude Agent SDK."""
import base64
import os
from pathlib import Path
from typing import Any, Dict, Optional

import aiofiles
import httpx

from shared.utils.logger import create_logger
from claude_agent_sdk import tool

logger = create_logger("RecognizeImage")


# Supported image formats
SUPPORTED_FORMATS = {".png", ".jpg", ".jpeg", ".gif", ".webp"}
MAX_FILE_SIZE = 5 * 1024 * 1024  # 5MB


def create_recognize_image_tool(
    api_key: str,
    base_url: Optional[str] = None,
    model: Optional[str] = None
):
    """Create RecognizeImage tool with given API configuration.
    
    Args:
        api_key: Anthropic API key
        base_url: Optional base URL for API
        model: Optional model name
    
    Returns:
        RecognizeImage tool instance
    """
    
    @tool(
        "RecognizeImage",
        """识别和分析图片内容。当需要查看、理解或分析图片文件时使用此工具。

支持的格式: PNG, JPG, JPEG, GIF, WEBP
最大文件大小: 5MB

使用场景:
- 识别验证码
- 分析截图内容
- 理解图表和数据可视化
- 提取图片中的文字
- 描述图片内容""",
        {
            "imagePath": str,  # 图片文件的绝对路径，例如: /path/to/image.png
            "question": str,   # 关于图片的问题或需要分析的内容(可选)
        }
    )
    async def recognize_image_impl(args: Dict[str, Any]) -> Dict[str, Any]:
        """Recognize and analyze image content.
        
        Args:
            args: Tool arguments containing imagePath and optional question
        
        Returns:
            Tool result
        """
        image_path = args.get("imagePath")
        question = args.get("question", "请详细描述这张图片的内容")
        
        if not image_path:
            return {
                "content": [{
                    "type": "text",
                    "text": "Error: imagePath is required"
                }],
                "isError": True
            }
        
        try:
            # Expand ~ to home directory
            expanded_path = os.path.expanduser(image_path)
            file_path = Path(expanded_path)
            
            # Validate file exists
            if not file_path.exists():
                return {
                    "content": [{
                        "type": "text",
                        "text": f"Error: Image file not found: {image_path}"
                    }],
                    "isError": True
                }
            
            # Validate file extension
            if file_path.suffix.lower() not in SUPPORTED_FORMATS:
                return {
                    "content": [{
                        "type": "text",
                        "text": f"Error: Unsupported image format. Supported formats: {', '.join(SUPPORTED_FORMATS)}"
                    }],
                    "isError": True
                }
            
            # Validate file size
            file_size = file_path.stat().st_size
            if file_size > MAX_FILE_SIZE:
                return {
                    "content": [{
                        "type": "text",
                        "text": f"Error: Image file too large ({file_size / 1024 / 1024:.2f}MB). Maximum size: 5MB"
                    }],
                    "isError": True
                }
            
            # Read image file
            async with aiofiles.open(file_path, "rb") as f:
                image_data = await f.read()
            
            # Determine media type
            media_type_map = {
                ".png": "image/png",
                ".jpg": "image/jpeg",
                ".jpeg": "image/jpeg",
                ".gif": "image/gif",
                ".webp": "image/webp",
            }
            media_type = media_type_map.get(file_path.suffix.lower(), "image/jpeg")
            
            # Encode image to base64
            image_base64 = base64.b64encode(image_data).decode("utf-8")
            
            # Use provided API configuration
            api_base_url = base_url or "https://api.anthropic.com"
            api_model = model or "claude-sonnet-4-20250514"
            
            # Ensure base_url ends with /v1/messages
            if not api_base_url.endswith("/messages"):
                if api_base_url.endswith("/v1"):
                    api_base_url = f"{api_base_url}/messages"
                elif not api_base_url.endswith("/"):
                    api_base_url = f"{api_base_url}/v1/messages"
            
            logger.info(f"[RecognizeImage] Using model: {api_model}, base_url: {api_base_url}")
            
            # Call Vision API
            async with httpx.AsyncClient(timeout=60.0) as client:
                response = await client.post(
                    api_base_url,
                    headers={
                        "x-api-key": api_key,
                        "anthropic-version": "2023-06-01",
                        "content-type": "application/json",
                    },
                    json={
                        "model": api_model,
                        "max_tokens": 2048,
                        "messages": [
                            {
                                "role": "user",
                                "content": [
                                    {
                                        "type": "image",
                                        "source": {
                                            "type": "base64",
                                            "media_type": media_type,
                                            "data": image_base64,
                                        },
                                    },
                                    {
                                        "type": "text",
                                        "text": question,
                                    },
                                ],
                            }
                        ],
                    },
                )
            
            if not response.is_success:
                error_text = response.text
                logger.error(f"[RecognizeImage] API error: {response.status_code} - {error_text}")
                return {
                    "content": [{
                        "type": "text",
                        "text": f"识别失败: {error_text}"
                    }],
                    "isError": True
                }
            
            # Parse response
            result = response.json()
            
            if "error" in result:
                return {
                    "content": [{
                        "type": "text",
                        "text": f"识别失败: {result['error'].get('message', 'Unknown error')}"
                    }],
                    "isError": True
                }
            
            # Extract text from response
            text = "无法识别图片内容"
            if "content" in result and result["content"]:
                for content_block in result["content"]:
                    if content_block.get("type") == "text":
                        text = content_block.get("text", text)
                        break
            
            logger.info(f"[RecognizeImage] Recognition successful, result length: {len(text)}")
            
            return {
                "content": [{
                    "type": "text",
                    "text": text
                }]
            }
            
        except Exception as e:
            logger.error(f"[RecognizeImage] Error: {e}")
            return {
                "content": [{
                    "type": "text",
                    "text": f"图片识别失败: {str(e)}"
                }],
                "isError": True
            }
    
    return recognize_image_impl
